### DISCUSSION TITLE
RFC Tracking Discussion: The Graph Type System

### DISCUSSION THREAD

**Discussion by Alfred-Mountfield** [blockprotocol/blockprotocol#418] [RFC]
This tracks the implementation of #352. Fixes and changes to the phrasing/correctness of the RFC should be submitted as PRs [to the RFC](https://github.com/blockprotocol/blockprotocol/blob/main/rfcs/text/0352-graph-type-system.md).

**Comment by thehabbos007** [comment]
@benwerner01 noticed that we are using `property-type` instead of `link-type` for the links in the reference-level explanation, for example [here](https://github.com/blockprotocol/blockprotocol/blob/main/rfcs/text/0352-graph-type-system.md#defining-relationships-between-entities-1). This seems to be an issue with almost all of the Entity Type schema links.

```json
"links": {
  "https://blockprotocol.org/types/@alice/property-type/written-by": {}
}
```
should be

```json
"links": {
  "https://blockprotocol.org/types/@alice/link-type/written-by": {}
}
```

**Reply by benwerner01** [reply]
Thanks for spotting, have opened [a quick PR](https://github.com/blockprotocol/blockprotocol/pull/597) fixing these mistakes in the RFC

**Comment by Alfred-Mountfield** [comment]
We are currently experimenting with having a single source of truth for Type System definitions. #503 is the start of an attempt to create a package that can create TypeScript and Rust definitions for the Type System elements, as well as having a single implementation of their validation logic.

If this is successful then we can look at using https://github.com/GREsau/schemars to generate the meta-schemas from the structs, which we could then use in the spec. As such for the time being, any inconsistencies or mistakes found in the original RFC are going to be implicitly fixed within the structs.

**Comment by indietyp** [comment]
I have been looking over the type system and have found the new `title_plural` for objects. I wanted to create a discussion for this as I have some problems/thoughts about it.

**Reply by indietyp** [reply]
Languages are hard. My problem is that titles may be localized, e.g. a German institution/company might create titles in German but share them.

Now you suddenly have a property type: `age` with the title: "Alter". That's awkward. 

The other thing is that some other languages (Arabic, for example) do not necessarily have the concept of only singular and plural.

**Reply by indietyp** [reply]
Now I completely understand that i18n is a super hard problem, my thoughts immediately went to a deviation of what [i18next](https://www.i18next.com/) uses.

```json
"title": {
  "en": {
    "0": "..."
    "1": "..."
    "other": "..."
  },
  "de": {}
}
```

which would use ISO 639 / RFC 5646 language codes

or as an example

```json
{
  "title": {
    "en": {
      "1": "Elevator",
      "other": "Elevators"
    },
    "en-uk": {
      "1": "Lift",
      "other": "Lifts"
    },
    "de": {
      "1": "Fahrstuhl",
      "other": "Fahrstühle"
    }
  }
}
```

As @Alfred-Mountfield pointed out this has one big problem: cognitive load and documents get large, fast.

**Reply by indietyp** [reply]
My other worry is: What if I want to change the title (I discovered that I made a typo for example, or I want to contribute another language), do I need to create a new schema version for that? What happens to all my old objects then?

**Reply by indietyp** [reply]
My other thought was to create something similar to:

```json
{
	"title": "reference in the graph"
}
```

This has the bonus that the title is just another link (tho idk about versioning), but comes with other overhead, e.g. when querying.

**Comment by indietyp** [comment]
While looking through the RFC, I noticed that entity types have a global default but no defaults for their individual properties.

I wonder if that makes sense. Is there a value to being able to set a default entity as opposed to set defaults for the properties?

imo it would be more logical and easier to implement defaults on properties instead. One can create a default entity with default property values or partially initialize one. Currently, one can initialize a default entity and not partially initialize entities with a global default.

I feel that default values on properties would make more sense as they are more flexible, more accessible to reason about, and already the standard in many tools (e.g. ORMs like SQLAlchemy or Prisma)

**Reply by Alfred-Mountfield** [reply]
Thanks for spotting! We definitely intended to have default values on a per-property basis (as in allowing an entity type to say it's default value for a property is "X"). That needs to be fixed and slipped through when designing the meta-schemas.

With regards to a general "here is a default entity", I too am not certain of the value of it. I think I agree with you that if we have default values for properties, that's probably enough.
